crates/boojum/src/gadgets/tower_extension/fq6.rs

```diff
@@ -0,0 +1,882 @@
+    }
+
+    /// Returns true if the `Fq6` element is zero.
+    pub fn is_zero<CS>(&mut self, cs: &mut CS) -> Boolean<F>
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let is_c0_zero = self.c0.is_zero(cs);
+        let is_c1_zero = self.c1.is_zero(cs);
+        let is_c2_zero = self.c2.is_zero(cs);
+        Boolean::multi_and(cs, &[is_c0_zero, is_c1_zero, is_c2_zero])
+    }
+
+    /// Adds two elements of `Fq6` by adding their components elementwise.
+    #[must_use]
+    pub fn add<CS>(&mut self, cs: &mut CS, other: &mut Self) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let c0 = self.c0.add(cs, &mut other.c0);
+        let c1 = self.c1.add(cs, &mut other.c1);
+        let c2 = self.c2.add(cs, &mut other.c2);
+        Self::new(c0, c1, c2)
+    }
+
+    /// Doubles the element of `Fq6` by doubling its components.
+    #[must_use]
+    pub fn double<CS>(&mut self, cs: &mut CS) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let c0 = self.c0.double(cs);
+        let c1 = self.c1.double(cs);
+        let c2 = self.c2.double(cs);
+        Self::new(c0, c1, c2)
+    }
+
+    /// Negates the element of `Fq6` by negating its components.
+    #[must_use]
+    pub fn negated<CS>(&mut self, cs: &mut CS) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let c0 = self.c0.negated(cs);
+        let c1 = self.c1.negated(cs);
+        let c2 = self.c2.negated(cs);
+        Self::new(c0, c1, c2)
+    }
+
+    /// Subtracts two elements of `Fq6` by subtracting their components elementwise.
+    #[must_use]
+    pub fn sub<CS>(&mut self, cs: &mut CS, other: &mut Self) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let c0 = self.c0.sub(cs, &mut other.c0);
+        let c1 = self.c1.sub(cs, &mut other.c1);
+        let c2 = self.c2.sub(cs, &mut other.c2);
+        Self::new(c0, c1, c2)
+    }
+
+    /// Multiplies the element in `Fq6` by a non-residue `v`.
+    pub fn mul_by_nonresidue<CS>(&mut self, cs: &mut CS) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        // c0, c1, c2 -> c2, c0, c1
+        let new_c2 = self.c2.mul_by_nonresidue(cs);
+        Self::new(new_c2, self.c0.clone(), self.c1.clone())
+    }
+
+    /// Multiplies the element in `Fq6` by a non-residue `\xi=9+u`.
+    pub fn mul_by_xi<CS>(&mut self, cs: &mut CS) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let new_c0 = self.c0.mul_by_nonresidue(cs);
+        let new_c1 = self.c1.mul_by_nonresidue(cs);
+        let new_c2 = self.c2.mul_by_nonresidue(cs);
+
+        Self::new(new_c0, new_c1, new_c2)
+    }
+
+    /// Multiplies two elements `a=a0+a1*v+a2*v^2`
+    /// and `b=b0+b1*v+b2*v^2` in `Fq6` using Karatsuba multiplication.
+    #[must_use]
+    pub fn mul<CS>(&mut self, cs: &mut CS, other: &mut Self) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let mut v0 = self.c0.mul(cs, &mut other.c0);
```

Comment:
I read an article you attached, and they claim that Toom Cook will be faster. 
Even with Karatsuba, there is too much multiplication

Reply:
Toom Cook may indeed be faster, but this is more relevant for higher-degree extensions, where the overhead of interpolation diminishes compared to the multiplications saved. To strengthen my argument, I also checked [go-ethereum](https://github.com/ethereum/go-ethereum/blob/master/crypto/bn256/google/gfp6.go#L118) and [franklin-crypto](https://github.com/matter-labs/franklin-crypto/blob/finalization/src/plonk/circuit/bigint/fp6.rs#L320) which as it seems both use the Karatsuba method